doc/development/tutorials/examples/recipe.py

```diff
@@ -0,0 +1,239 @@
+            objlis = imap[ingredient]
+            for objname in objlis:
+                dispname, typ, docname, anchor = objs[objname]
+                lis.append((
+                    dispname, 0, docname,
+                    anchor,
+                    docname, '', typ
+                ))
+        re = [(k, v) for k, v in sorted(content.items())]
+
+        return (re, True)
+
+
+class RecipeIndex(Index):
+    name = 'rcp'
+    localname = 'Recipe Index'
+    shortname = 'Recipe'
+
+    def __init__(self, *args, **kwargs):
+        super(RecipeIndex, self).__init__(*args, **kwargs)
+
+    def generate(self, docnames=None):
+        """Return entries for the index given by *name*.
+
+        If *docnames* is given, restrict to entries referring to these
+        docnames.  The return value is a tuple of ``(content, collapse)``,
+        where:
+
+        *collapse* is a boolean that determines if sub-entries should
+        start collapsed (for output formats that support collapsing
+        sub-entries).
+
+        *content* is a sequence of ``(letter, entries)`` tuples, where *letter*
+        is the "heading" for the given *entries*, usually the starting letter.
+
+        *entries* is a sequence of single entries, where a single entry is a
+        sequence ``[name, subtype, docname, anchor, extra, qualifier, descr]``.
+
+        The items in this sequence have the following meaning:
+
+        - `name` -- the name of the index entry to be displayed
+        - `subtype` -- sub-entry related type:
+          - ``0`` -- normal entry
+          - ``1`` -- entry with sub-entries
+          - ``2`` -- sub-entry
+        - `docname` -- docname where the entry is located
+        - `anchor` -- anchor for the entry within `docname`
+        - `extra` -- extra info for the entry
+        - `qualifier` -- qualifier for the description
+        - `descr` -- description for the entry
+
+        Qualifier and description are not rendered by some builders, such as
+        the LaTeX builder.
+        """
+
+        content = {}
+        items = ((name, dispname, typ, docname, anchor)
+                 for name, dispname, typ, docname, anchor, prio
+                 in self.domain.get_objects())
+        items = sorted(items, key=lambda item: item[0])
+        for name, dispname, typ, docname, anchor in items:
+            lis = content.setdefault('Recipe', [])
+            lis.append((
+                dispname, 0, docname,
+                anchor,
+                docname, '', typ
+            ))
+        re = [(k, v) for k, v in sorted(content.items())]
+
+        return (re, True)
+
+
+class RecipeDomain(Domain):
+
+    name = 'rcp'
+    label = 'Recipe Sample'
+
+    roles = {
+        'reref': XRefRole()
+    }
+
+    directives = {
+        'recipe': RecipeDirective,
+    }
+
+    indices = {
+        RecipeIndex,
+        IngredientIndex
+    }
+
+    initial_data = {
+        'objects': [],  # object list
```

Comment:
Nobody restricts the name of keys. So it would be easy to understand if we rename this to `recipes` key.

Reply:
+1